sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
     std::unique_ptr<sstable_set_impl> _impl;
-    schema_ptr _schema;
+    std::unordered_map<utils::UUID, std::unordered_set<shared_sstable>> _all_runs;
+    std::unordered_map<shared_sstable, unsigned> _times_added;
+    std::set<shared_sstable> _all;
+public:
+    versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl);
+    void insert(shared_sstable sst);
+    void erase(shared_sstable sst);
+    std::vector<shared_sstable> select(const dht::partition_range& range) const;
+    std::unordered_set<shared_sstable> select_by_run_id(utils::UUID run_id) const;
+    const std::set<shared_sstable>& all() const;
+    void remove(shared_sstable sst);
+    std::unique_ptr<incremental_selector_impl> make_incremental_selector();
+};
+
+class sstable_set_version;
+
+class sstable_set_version_reference {
+    mutable sstable_set_version* _p = nullptr;
+public:
+    ~sstable_set_version_reference();
+    sstable_set_version_reference() = default;
+    sstable_set_version_reference(sstable_set_version* p);
```

Comment:
This should be private if possible. I'm guessing this makes some assumptions about reference counting, so if it's used accidentally it can cause damage.

Reply:
(and explicit)

Reply:
I've changed it to private and explicit